text/3127-trim-paths.md

```diff
@@ -0,0 +1,255 @@
+- `2` or `true`: sanitise paths in all compilation outputs, including compiled executable/library, debug information, and compiler diagnostics.
+
+The default release profile uses option `1`. You can also manually override it by specifying this option in `Cargo.toml`:
+```toml
+[profile.dev]
+trim-paths = 2
+
+[profile.release]
+trim-paths = 0
+```
+
+When a path is in scope for sanitisation, it is handled by the following rules:
+
+1. Path to the source files of the standard and core library (sysroot) will begin with `/rustc/[rustc commit hash]`.
+   E.g. `/home/username/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs` -> 
+   `/rustc/fe72845f7bb6a77b9e671e6a4f32fe714962cec4/library/core/src/result.rs`
+2. Path to the working directory will be stripped. E.g. `/home/username/crate/src/lib.rs` -> `src/lib.rs`.
+3. Path to packages outside of the working directory will be replaced with `[package name]-[version]`. E.g. `/home/username/deps/foo/src/lib.rs` -> `foo-0.1.0/src/lib.rs`
+
+When a path to the source files of the standard and core library is *not* in scope for sanitisation, the emitted path will depend on if `rust-src` component
+is present. If it is, then the real path pointing to a copy of the source files on your file system will be emitted; if it isn't, then they will
+show up as `/rustc/[rustc commit hash]/library/...` (just like when it is selected for sanitisation). Paths to all other source files will not be affected.
+
+This will not affect any hard-coded paths in the source code.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `trim-paths` implementation in Cargo
+
+We only need to change the behaviour for `Test` and `Build` compile modes. 
+
+If `trim-paths` is `0` (`false`), no extra flag is supplied to `rustc`.
+
+If `trim-paths` is `1` or `2` (`true`), then two `--remap-path-prefix` arguments are supplied to `rustc`:
+- From the path of the local sysroot to `/rustc/[commit hash]`. 
+- If the compilation unit is under the working directory, from the the working directory absolute path to empty string.
+  If it's outside the working directory, from the absolute path of the package root to `[package name]-[package version]`.
+
+A further `--remap-path-scope` is also supplied for options `1` and `2`:
+
+If `trim-path` is `1`, then it depends on the setting of `split-debuginfo` (whether the setting is explicitly supplied or from the default)
+- If `split-debuginfo` is `off`, then `--remap-path-scope=macro,debuginfo`.
+- If `split-debuginfo` is `packed` or `unpacked`, then `--remap-path-scope=macro,split-debuginfo-file`
+
+This is because we always want to remap panic messages as they will always be embedded in executable/library. We need to sanitise debug information
+if they are embedded, but don't need to touch them if they are split. However in case they are split we need to sanitise the paths to these split files 
```

Comment:
Rather than having to special-case whether split-debuginfo is on or off, could we just have some additional separate values and use those? If we add an `unsplit-debuginfo` scope that means "remap paths in debuginfo that's included in the compiled file, but not debuginfo that's split out", we can enable that unconditionally in level 1. And if we define `split-debuginfo-path` as being a no-op if there's no split debuginfo, we can enable that unconditionally in level 1 as well.

Reply:
`unsplit-debuginfo` is a good idea, added.